### PR TITLE
Added teleportation example

### PR DESCRIPTION
**Pull request by AbeerVaishnav13**
Hey @tgehr ,
Silq is a really great language and I'm really addicted to it now (haven't used qiskit or Cirq for 3 weeks straight).

Added an example for Quantum Teleportation over multiple message qubits. Hope this gets merged :)

Thanks
Abeer

### CODE REVIEW

File: test/teleportation.slq [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+def createBellPair[n:!ℕ]():uint[n]^2 {
+	// Create qureg0 in superposition
+	x := 0:uint[n];
+	for i in [0..n) { x[i] := H(x[i]); }
+
+	// Perform a CNOT on each qubit of qureg1
+	y := 0:uint[n];
+	for i in [0..n) {
+		if x[i] { y[i] := X(y[i]); }
+	}
```
Comment (tgehr) [inline test/teleportation.slq:32]: This is of course a valid way to do it, but note that you can just write `y := dup(x);` for the same overall effect.

Comment (AbeerVaishnav13) [reply test/teleportation.slq:32]: Well isn't that more like entangling the states to form a bell pair/ghz state? I'm not sure how is that can be seen as teleportation?

**Comment by AbeerVaishnav13** [comment]
Thanks for meeting the pull request @tgehr !

File: test/teleportation.slq (hunk 1 above)
Comment (ZoeDreams) [reply test/teleportation.slq:32]: Great example; however please correct me, but isn't this is an example of an implementation of a Hadamard Gate -> more info https://en.wikipedia.org/wiki/Quantum_logic_gate#Hadamard_(H)_gate

Teleportation requires creating an array with an entangled state; then separating the array while in an entangled state, creating a separate quantum circuit. See this recent article for more information https://www.sciencealert.com/physicists-have-teleported-information-between-particles-of-matter-for-the-first-time

Comment (ZoeDreams) [reply test/teleportation.slq:32]: > Well isn't that more like entangling the states to form a bell pair/ghz state? I'm not sure how is that can be seen as teleportation?

i agree.

Comment (AbeerVaishnav13) [reply test/teleportation.slq:32]: Re:
> Great example; however please correct me, but isn't this is an example of an implementation of a Hadamard Gate -> more info https://en.wikipedia.org/wiki/Quantum_logic_gate#Hadamard_(H)_gate
> 
> Teleportation requires creating an array with an entangled state; then separating the array while in an entangled state, creating a separate quantum circuit. See this recent article for more information https://www.sciencealert.com/physicists-have-teleported-information-between-particles-of-matter-for-the-first-time

Hi @ZoeDreams,
This is not an example of the implementation of the Hadamard gate. The H gate is used to create an equal superposition of one or more states something like this: H|0⟩ = 1/√2⋅(|0⟩ + |1⟩) and H|1⟩ = 1/√2⋅(|0⟩ - |1⟩). A similar thing can be extended to multiple qubits.

Yes, you're right. Teleportation requires creating an array with an entangled state; then separating the array while in an entangled state, creating a separate quantum circuit. And that's what is happening here. Do remember that quantum teleportation is also an algorithm and it is implemented here as an 'algorithm'. The function `createBellPair()` returns the two entangled registers `x` and `y`, which are then separated by passing one of them to the `alice()` function and another to the `bob()` function. The link which you mention is just the implementation of teleportation on two different physical processors, which is slightly different from what is being done here.

Here, teleportation is shown between two 'quregisters' belonging to the **same quantum system**. Whereas, in the experiment which has been mentioned by you, these two quregisters (electrons) exist as **two different quantum particles** within the same quantum system. Also, the communication channel between Alice and Bob is simulated by the logic that the output of `alice()` function is passed to `bob()` function, which then returns the teleported state.

One more thing is that this code just provides an algorithmic approach of Quantum Teleportation. The `main()` function will be the one to prepare the message quantum state to be teleported and also communicate the measured classical bits from Alice to Bob. That can be implemented as given [here](https://github.com/AbeerVaishnav13/Quantum-Algorithms-in-Silq/blob/master/Quantum-Algorithms/Teleportation.slq).

**Comment by AbeerVaishnav13** [comment]
Hi @tgehr,
I don't see an example for Shor's algorithm in the silq repo. Is there a specific reason for not including it or can I see it as an opportunity to contribute one more example into silq? 😅

**Comment by tgehr** [comment]
@AbeerVaishnav13 Well, there is Shor's algorithm for discrete logarithm: https://github.com/eth-sri/silq/blob/master/test/dlog.slq

**Comment by AbeerVaishnav13** [comment]
Ohh yes @tgehr I meant Shor's Algorithm for factoring. My bad.

Ok sure, I'll create another pull request shortly for  the factoring algorithm.

Thanks :)

File: test/teleportation.slq (hunk 1 above)
Comment (ZoeDreams) [reply test/teleportation.slq:32]: Thank you for your reply and clarification. if i understand you in clear terms this is teleporting registers (pointers) to entangled bits in the same array? Again thank you for taking the time to respond clearly.

**Pull request merged**